### PR TITLE
Fix post-restart powershell failure

### LINKED ISSUE #4994: Versions above 0.11.0 fail reproducably when provisioning Windows AMIs in AWS
After upgrading Packer from 0.11.0 to any newer version (incl. 1.0.0), it starts to fail reproducably when creating Windows Server 2008R2 AMIs from Amazon base AMIs in AWS. Windows Server 2012R2 AMIs are still provisioned fine. The last bits of log output are:
```
[2017-06-09T12:43:01] INFO  -- axc : ==> amazon-ebs: Restarting Machine
[2017-06-09T12:43:01] INFO  -- axc : ==> amazon-ebs: Waiting for machine to restart...
[2017-06-09T12:43:01] INFO  -- axc :     amazon-ebs: A system shutdown is in progress.(1115)
[2017-06-09T12:43:53] INFO  -- axc :     amazon-ebs: WIN-2E19G4EGGMK restarted.
[2017-06-09T12:43:53] INFO  -- axc : ==> amazon-ebs: Machine successfully restarted, moving on
[2017-06-09T12:43:53] INFO  -- axc : ==> amazon-ebs: Provisioning with Powershell...
[2017-06-09T12:43:53] INFO  -- axc : ==> amazon-ebs: Provisioning with shell script: /path/to/install_puppet.ps1
[2017-06-09T12:44:24] INFO  -- axc : ==> amazon-ebs: Terminating the source AWS instance...
[2017-06-09T12:44:55] INFO  -- axc : ==> amazon-ebs: Cleaning up any extra volumes...
[2017-06-09T12:44:55] INFO  -- axc : ==> amazon-ebs: No volumes to clean up, skipping
[2017-06-09T12:44:55] INFO  -- axc : ==> amazon-ebs: Deleting temporary keypair...
[2017-06-09T12:44:55] INFO  -- axc : Build 'amazon-ebs' errored: Script exited with non-zero exit status: 2147942487. Allowed exit codes are: [0]
[2017-06-09T12:44:55] INFO  -- axc :
[2017-06-09T12:44:55] INFO  -- axc : ==> Some builds didn't complete successfully and had errors:
[2017-06-09T12:44:55] INFO  -- axc : --> amazon-ebs: Script exited with non-zero exit status: 2147942487. Allowed exit codes are: [0]
[2017-06-09T12:44:55] INFO  -- axc :
[2017-06-09T12:44:55] INFO  -- axc : ==> Builds finished but no artifacts were created.
```

- Versions: 0.12.x, 1.0.0
- Ubuntu Linux
- https://gist.github.com/dhs-rec/a5cf71c78ba7f8a832ab0fcb8d49483e

### LINKED ISSUE #5064: First powershell script after installing powershell 5 and rebooting doesn't run
I've got a weird situation with packer `1.0.2`, which makes it not run a script after powershell 5 is installed and the guest is rebooted. Subsequent files are executed okay.

In the output, it should have `Installing Microsoft Windows SDK for Windows 10 10.0.26624`, and also should actually install the `windows-sdk-10.0` from chocolatey. However, it does not.

The log says there's 0 bytes returned on STDERR and STDOUT.

When I log onto the box, and run `c:\windows\temp\script.ps1`, it works fine - file has uploaded okay.

If my packer file has the same file listed twice it works:
```
      scripts": [
        "./scripts/install-win10-sdk.ps1"
        "./scripts/install-win10-sdk.ps1"
      ],
```

- Packer version from `packer version`
- Host platform: `macOS Sierra 10.12.5 (16F73)`

templates, scripts & logs: https://gist.github.com/matt-richardson/bee5b3fb7de5f5c0b78cc92884dd27e4

### PR DESCRIPTION
**Pull request by SwampDragons**
Two fixes:  

1) There was a bug in powershell where if we ran one powershell script and then another, both were uploaded to the same remote path and the first script was cached somehow.  This meant that when the time came to run the second script, we'd actually be trying to rerun the first.  

This code change circumvents that problem by making sure every script uploaded is sent to a unique remote path (note that all of this is overridden if the user sets the "remote_path" option in the packer config)

2) Windows restart provisioner was not waiting for the computer to come fully online.  It'd check that the communicator could run an "echo <computername> restarted" command but wouldn't make sure the echo actually succeeded.  This led to the provisioner returning too soon, and the first powershell provisioner run post-reboot to fail.  I've modified the code to actually parse stdout and make sure the echo suceeded.

Closes #5064
Closes #4994

### CODE REVIEW

**Comment by SwampDragons** [comment]
Ah, need to update tests before merging 😬

**Comment by SwampDragons** [comment]
TestProvision_waitForCommunicator is failing.  I'm gonna need to figure out how to rework the test.

File: provisioner/windows-restart/provisioner.go [hunk 1]
```diff
@@ -183,12 +189,20 @@ var waitForCommunicator = func(p *Provisioner) error {
...
 		err := cmd.StartWithUi(p.comm, p.ui)
+
 		if err != nil {
 			log.Printf("Communication connection err: %s", err)
 			continue
 		}
 
 		log.Printf("Connected to machine")
+		stdoutToRead := buf2.String()
+		buf2.Reset()
```
Comment (mwhooker) [inline provisioner/windows-restart/provisioner.go]: I don't _think_ these resets are needed since the buf vars get redeclared at the top of the loop

**Comment by matt-richardson** [comment]
@SwampDragons - will this fix work with the "waiting for winrm" as well? I've got an intermittent failure where it uploads a file not long after connecting to winrm - i wonder if its hitting the same issue...?

Actually, looking at https://github.com/hashicorp/packer/blob/master/helper/communicator/step_connect_winrm.go#L126, it seems there's a strong chance this issue could happen there too...

**Comment by SwampDragons** [comment]
I think I see what you're saying.  We might be able to add another echo-test here to make sure winrm is ready to roll before moving forwards with the upload.

File: provisioner/windows-restart/provisioner.go (hunk 1 above)
Comment (SwampDragons) [reply provisioner/windows-restart/provisioner.go]: ah, you're probably right.  I initially had the buffer variables declared outside the loop.

**Comment by matt-richardson** [comment]
Just to make sure we're on the same page - I'm thinking the echo test will probably be required in the [waitForWinRM](https://github.com/hashicorp/packer/blob/master/helper/communicator/step_connect_winrm.go#L84) function, before it declares that WinRM is available, not in the file upload routine.

**Comment by matt-richardson** [comment]
@SwampDragons - I've just run three builds in a row with a new `shell-local` provisioner that waits 60 seconds after the `Waiting for WinRM` before trying to upload the files, and they all passed. Feels reasonably convincing that an echo test will help in this circumstance will help.

**Comment by SwampDragons** [comment]
Awesome!  I'll probably not get to coding it up this milestone, since I'm likely gonna be reading PRs til my eyes bleed between now and the 1.1.0 release -- but I'll try, or you can make a PR and I'll do my best to slot it in.

**Comment by matt-richardson** [comment]
Want me to create a new issue for that one, so we don't lose track?

**Comment by SwampDragons** [comment]
Ah, yeah, sorry, I meant to do it yesterday but I forgot >.<

**Comment by SwampDragons** [comment]
@matt-richardson if you follow the trail of references above you'll end up at a patch that should do this -- mind running a few builds and telling me if you see improved stability?

**Pull request merged**